compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -4333,7 +4338,9 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
             if needsTupledDual(ref, pt) && Feature.autoTuplingEnabled =>
               adapt(tree, pt.tupledDual, locked)
             case _ =>
-              adaptOverloaded(ref)
+              val ctorResultType = applyProxy(tree)
```

Comment:
This happens three times. Suggestion:

Make newExpr like this:
```scala
    /** If `tree` is a constructor proxy reference, convert it to a `new` expression,
     *  otherwise return EmptyTree 
     */
    def newExpr(tree: Tree) = ...
```
then you can use `orElse` on trees for the fallback. I.e.
```scala
    newExpr(ref).orElse(applyOverloaded(ref))
```